### PR TITLE
209 ES adaptor stability

### LINKED ISSUE #209: Elasticsearch adaptor stability
the current version is out of date with all of the breaking API changes that have been made between versions 1.X -> 2.X -> 5.X

### PR DESCRIPTION
**Pull request by jipperinbham**
- [x] CHANGELOG.md updated
- [x] README.md updated

switch from `elastigo` to `elastic` library and create underlying clients for each major version of elasticsearch.

fixes #209, #222, #167, and #159

### CODE REVIEW

File: pkg/adaptor/elasticsearch/clients/registry.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package clients
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/registry.go]: I can see us having/wanting to use this sort of registry for other db types outside of ES. Let's make an issue for making this generic and have the other adaptors use the registry as well. I guess we would need to be able to get the version information through an API call like we do for ES. Should be possible.

File: pkg/adaptor/elasticsearch/clients/v5/writer.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	return w
+}
+
+func (w *Writer) Write(msg message.Msg) func(client.Session) error {
+	return func(s client.Session) error {
+		i, t, _ := message.SplitNamespace(msg)
+		var id string
+		if _, ok := msg.Data()["_id"]; ok {
+			id = msg.ID()
+			msg.Data().Delete("_id")
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v5/writer.go]: Will we have to restore the `id` field after the write for nodes further down in the pipeline that may expect `_id` to be present? I'm not sure if `msg.Data()` returns a copy or not. I think not.

File: pkg/adaptor/elasticsearch/clients/v5/writer.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		case ops.Update:
+			br = elastic.NewBulkUpdateRequest().Index(i).Type(t).Id(id).Doc(msg.Data())
+		}
+		w.bp.Add(br)
+		return nil
+	}
+}
+
+func (w *Writer) Close() {
+	log.Infoln("flushing BulkProcessor")
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v5/writer.go]: Would be nice to have this be a little more descriptive (that it's an ES bulk processor)

File: pkg/adaptor/elasticsearch/elasticsearch.go [hunk 4]
```diff
@@ -77,69 +134,124 @@ func init() {
...
 }
 
 // Stop the adaptor
 func (e *Elasticsearch) Stop() error {
-	if e.running {
-		e.running = false
-		e.pipe.Stop()
-		e.indexer.Stop()
-	}
+	log.With("path", e.path).Infoln("adaptor Stopping...")
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/elasticsearch.go]: More descriptive would be nice here as well

File: pkg/adaptor/elasticsearch/elasticsearch_test.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+	"github.com/compose/transporter/pkg/adaptor"
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/ops"
+	"github.com/compose/transporter/pkg/pipe"
+)
+
+const (
+	DefaultURL = "http://127.0.0.1:9200"
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/elasticsearch_test.go]: This will get exposed by the elasticsearch package, better to be private

File: pkg/adaptor/elasticsearch/clients/v2/writer_test.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	"time"
+
+	"github.com/compose/transporter/pkg/adaptor/elasticsearch/clients"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/ops"
+)
+
+const (
+	DefaultURL   = "http://127.0.0.1:9200"
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v2/writer_test.go]: Better to be private

File: pkg/adaptor/elasticsearch/clients/v2/writer_test.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	DefaultIndex = "test_v2"
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v2/writer_test.go]: Better to be private

File: pkg/adaptor/elasticsearch/clients/v1/writer_test.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+		os.Exit(1)
+	}
+	code := m.Run()
+	shutdown()
+	os.Exit(code)
+}
+
+func shutdown() {
+	log.Debugln("shutting down tests")
+	// clearTestData()
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v1/writer_test.go]: should this be commented?

File: pkg/adaptor/elasticsearch/clients/v1/writer_test.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	log.Debugln("tests shutdown complete")
+}
+
+type CountResponse struct {
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v1/writer_test.go]: Make this private

File: pkg/adaptor/elasticsearch/clients/v2/writer.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	return w
+}
+
+func (w *Writer) Write(msg message.Msg) func(client.Session) error {
+	return func(s client.Session) error {
+		i, t, _ := message.SplitNamespace(msg)
+		var id string
+		if _, ok := msg.Data()["_id"]; ok {
+			id = msg.ID()
+			msg.Data().Delete("_id")
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v2/writer.go]: Same comment as with v1, do we need to restore?

File: pkg/adaptor/elasticsearch/clients/v2/writer.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
+package v2
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v2/writer.go]: I know you're aware of all the code duplication, just want to make sure we track it in an issue to investigate clever ways to do the versioning.

File: pkg/adaptor/elasticsearch/elasticsearch.go [hunk 12]
```diff
@@ -77,69 +134,124 @@ func init() {
...
+		httpClient = &http.Client{
+			Timeout: t,
 		}
 	}
 
-	// we might have a port in the host bit
-	hostBits := strings.Split(e.uri.Host, ":")
-	if len(hostBits) > 1 {
-		client.SetPort(hostBits[1])
+	for _, vc := range clients.Clients {
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/elasticsearch.go]: If we make the registry generic, I think a function to return the proper client given the adaptor type and version would be better than having this logic in the adaptor. Even for just elasticsearch, I think it would be better to have this in the registry.

File: pkg/log/logger.go [hunk 13]
```diff
@@ -96,6 +98,10 @@ func (l logger) Errorf(format string, args ...interface{}) {
 	l.entry.Errorf(format, args...)
 }
 
+func (l logger) Printf(format string, args ...interface{}) {
```
Comment (trinchan) [inline pkg/log/logger.go]: Why is this necessary if we have `logger.Infof`?

Comment (jipperinbham) [inline pkg/log/logger.go]: this is to fulfill the `Logger` interface in the new elasticsearch lib, https://github.com/olivere/elastic/wiki/Logging

File: pkg/adaptor/elasticsearch/clients/v2/writer.go (hunk 11 above)
Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/clients/v2/writer.go]: 👍

File: pkg/adaptor/elasticsearch/clients/v5/writer.go (hunk 2 above)
Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/clients/v5/writer.go]: `msg.Data()` does not return a copy and I hadn't really thought about the use case for chaining the adaptors in such a way, it does bring up the issue of whether the msg.Data should be immutable because in instances were multiple `Sink` adaptors are attached to the same `Source`, this code could cause a race condition with a concurrent map read/write.

File: pkg/adaptor/elasticsearch/clients/registry.go (hunk 1 above)
Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/clients/registry.go]: issue created to track that work, https://github.com/compose/transporter/issues/234

File: pkg/adaptor/elasticsearch/clients/registry.go [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+type VersionedClient struct {
+	Constraint version.Constraints
+	Creator    Creator
+}
+
+// Creator defines the func signature expected for any implementing client.Writer
+type Creator func(chan struct{}, *sync.WaitGroup, *ClientOptions) (client.Writer, error)
+
+// Clients contains the map of versioned clients
+var Clients = map[string]*VersionedClient{}
```
Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/clients/registry.go]: planning on addressing this in https://github.com/compose/transporter/issues/234

File: pkg/adaptor/elasticsearch/elasticsearch.go (hunk 4 above)
Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/elasticsearch.go]: suggestions?

File: pkg/adaptor/elasticsearch/clients/v2/writer.go (hunk 10 above)
Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/clients/v2/writer.go]: addressed by making a copy of the original `msg.Data()` in the adaptor.

File: pkg/adaptor/elasticsearch/clients/v5/writer.go (hunk 2 above)
Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/clients/v5/writer.go]: see https://github.com/compose/transporter/pull/233#discussion_r97595070

File: pkg/adaptor/elasticsearch/clients/v2/writer.go [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+		esClient, err := elastic.NewClient(esOptions...)
+		if err != nil {
+			return nil, err
+		}
+		return newWriter(esClient, done, wg), nil
+	})
+}
+
+func newWriter(client *elastic.Client, done chan struct{}, wg *sync.WaitGroup) *Writer {
+	p, _ := client.BulkProcessor().
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v2/writer.go]: Not sure what we'd do with the errors, but right now we're not catching any of the errors from the BulkProcessor. You can use a [BulkAfterFunc](https://github.com/olivere/elastic/blob/1368b91516f39100baf5e23a02ae9f090c0d90e2/bulk_processor.go#L70) and [.After(callback)](https://github.com/olivere/elastic/blob/1368b91516f39100baf5e23a02ae9f090c0d90e2/bulk_processor.go#L78) to get at the errors though. Would be bad to throw away errors without any output.

File: pkg/adaptor/elasticsearch/clients/v5/writer.go [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+func newWriter(client *elastic.Client, done chan struct{}, wg *sync.WaitGroup) *Writer {
+	p, _ := client.BulkProcessor().
+		Name("TransporterWorker-1").
+		Workers(2).
+		BulkActions(1000).               // commit if # requests >= 1000
+		BulkSize(2 << 20).               // commit if size of requests >= 2 MB
+		FlushInterval(30 * time.Second). // commit every 30s
+		Do()
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/clients/v5/writer.go]: Same here with the error handling

File: pkg/adaptor/elasticsearch/elasticsearch.go (hunk 4 above)
Comment (trinchan) [inline pkg/adaptor/elasticsearch/elasticsearch.go]: ah i noticed the e.path there, it will show the adaptor type, all good.

File: pkg/adaptor/elasticsearch/transport.go [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+// AWSTransport handles wrapping requests to AWS Elasticsearch service
+type AWSTransport struct {
+	Credentials awsauth.Credentials
+	transport   http.RoundTripper
+}
+
+func newTransport(accessKeyID, secretAccessKey string) http.RoundTripper {
+	t := http.DefaultTransport
+	if accessKeyID != "" && secretAccessKey != "" {
```
Comment (trinchan) [inline pkg/adaptor/elasticsearch/transport.go]: is there any way to test the creds and error sooner?

File: pkg/adaptor/elasticsearch/clients/v2/writer.go (hunk 15 above)
Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/clients/v2/writer.go]: ah, thanks for the reminder, I wanted to log any errors so I'll get that in now

File: pkg/adaptor/elasticsearch/transport.go (hunk 17 above)
Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/transport.go]: what would we be testing here? when adding support for AWS signed auth, this seemed like the best/cleanest route to go.

Comment (trinchan) [inline pkg/adaptor/elasticsearch/transport.go]: was just wondering if there was a way to bail out early if the creds were bad

Comment (jipperinbham) [inline pkg/adaptor/elasticsearch/transport.go]: gotcha, I honestly don't know
